### PR TITLE
Adding JSON model reader infrastructure

### PR DESCRIPTION
**Pull request by jleidel**
Adding JSON model reader infrastructure using PR #748 as the basis for the format.  This infrastructure allows users to specify model inputs using a standard JSON format.  

Supports
- components
- subcomponents
- recursive subcomponents (subcomponents of subcomponents)
- links
- program options

Does not support:
- statistics (forthcoming based upon the changes to the writer output)

### CODE REVIEW

File: src/sst/core/model/json/jsonmodel.h [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+#define SST_MODEL_JSON_COMPONENTID_IDX     1
+#define SST_MODEL_JSON_CONFIGCOMPONENT_IDX 2
+
+    typedef std::vector<std::tuple<std::string, ComponentId_t, ConfigComponent*>> CompTuple;
+    CompTuple                                                                     configGraphTuple;
+
+    void          recursiveSubcomponent(ConfigComponent* Parent, nlohmann::basic_json<> compArray);
+    void          discoverProgramOptions(json jFile);
+    void          discoverComponents(json jFile);
+    void          discoverLinks(json jFile);
```
Comment (calewis) [inline src/sst/core/model/json/jsonmodel.h:82]: Just one quick point before a bigger review, unless the json objects are shallow copy, can these be `const&` or do you use the `jFile` for temp storage in these functions?

File: src/sst/core/model/json/jsonmodel.cc [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+// Copyright (c) 2009-2021, NTESS
+// All rights reserved.
+//
+// This file is part of the SST software package. For license
+// information, see the LICENSE file in the top level directory of the
+// distribution.
+
+#include "sst/core/model/json/jsonmodel.h"
+
+#ifdef SST_CONFIG_HAVE_MPI
```
Comment (feldergast) [inline src/sst/core/model/json/jsonmodel.cc:16]: Is there are reason to include mpi in this file?  I don't see MPI used anywhere.

Comment (jleidel) [reply src/sst/core/model/json/jsonmodel.cc:16]: This was copied over from the python model reader, so I can remove it.

File: src/sst/core/model/json/jsonmodel.h (hunk 1 above)
Comment (jleidel) [reply src/sst/core/model/json/jsonmodel.h:82]: they can be const

Comment (jleidel) [reply src/sst/core/model/json/jsonmodel.h:82]: on second thought, no.  it causes an assertion inside nholman::json

File: src/sst/core/model/json/jsonmodel.cc (hunk 2 above)
Comment (jleidel) [reply src/sst/core/model/json/jsonmodel.cc]: removed in commit cc7003ef

**Review by feldergast — COMMENTED** [review]
Okay to test

**Review by feldergast — APPROVED** [review]

**Pull request merged**